src/Microsoft.AspNetCore.Mvc.Core/Internal/AuthorizationApplicationModelProvider.cs

```diff
@@ -36,21 +35,49 @@ public void OnProvidersExecuting(ApplicationModelProviderContext context)
 
             foreach (var controllerModel in context.Result.Controllers)
             {
-                controllerModel.Filters.Add(new AuthorizeFilter(_policyProvider, controllerModel.Attributes.OfType<IAuthorizeData>()));
+                var controllerModelAuthData = controllerModel.Attributes.OfType<IAuthorizeData>().ToArray();
+                if (controllerModelAuthData.Length > 0)
+                {
+                    var filter = GetFilter(controllerModelAuthData);
+                    controllerModel.Filters.Add(filter);
+                }
                 foreach (var attribute in controllerModel.Attributes.OfType<IAllowAnonymous>())
                 {
                     controllerModel.Filters.Add(new AllowAnonymousFilter());
                 }
 
                 foreach (var actionModel in controllerModel.Actions)
                 {
-                    actionModel.Filters.Add(new AuthorizeFilter(_policyProvider, actionModel.Attributes.OfType<IAuthorizeData>()));
+                    var actionModelAuthData = actionModel.Attributes.OfType<IAuthorizeData>().ToArray();
+                    if (actionModelAuthData.Length > 0)
+                    {
+                        var filter = GetFilter(actionModelAuthData);
+                        controllerModel.Filters.Add(filter);
+                    }
+
                     foreach (var attribute in actionModel.Attributes.OfType<IAllowAnonymous>())
                     {
                         actionModel.Filters.Add(new AllowAnonymousFilter());
                     }
                 }
             }
         }
+
+        private AuthorizeFilter GetFilter(IEnumerable<IAuthorizeData> authData)
+        {
+            // The default policy provider will make the same policy for given input, so only make it once.
+            AuthorizeFilter filter;
+            if (_policyProvider.GetType() == typeof(DefaultAuthorizationPolicyProvider))
+            {
+                var policy = AuthorizationPolicy.CombineAsync(_policyProvider, authData).Result;
```

Comment:
This should be `GetAwaiter().GetResult()` - it's subtle but there are differences in how exceptions are reported.